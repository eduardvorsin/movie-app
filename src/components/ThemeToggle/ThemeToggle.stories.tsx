import type { Meta, StoryObj } from '@storybook/react';
import ThemeToggle, { Props } from './ThemeToggle';
import { useTranslation } from 'react-i18next';

const meta: Meta<typeof ThemeToggle> = {
	title: 'UI/ThemeToggle',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: ThemeToggle,
	parameters: {
		docs: {
			description: {
				component: 'The ThemeToggle component is a visual interface tool designed to switch between different themes or display styles of a web application. This component provides users with the ability to choose the preferred appearance of the application.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof ThemeToggle>;

const ThemeToggleWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = { label: t('themeToggle.label') };

	return (
		<ThemeToggle
			{...props}
			dictionary={dictionary}
		/>
	);
}

export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (<ThemeToggleWithHooks {...args} />),
};