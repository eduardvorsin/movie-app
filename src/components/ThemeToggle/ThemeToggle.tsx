'use client';
import { ThemeContext } from '@/context/ThemeProvider/ThemeProvider';
import { useTranslation } from '@/i18n/client';
import { Locales, fallbackLng } from '@/i18n/settings';
import { useParams } from 'next/navigation';
import { ChangeEventHandler, useContext, useState } from 'react';
import Toggle from '@/components/Toggle/Toggle';
import MoonIcon from '../../assets/icons/moon.svg';
import SunIcon from '../../assets/icons/sun.svg';
import { GeneralProps } from '@/types/shared';

type Props = GeneralProps;

export default function ThemeToggle({
	className,
	testId,
	...props
}: Props) {
	const lang = useParams()?.lang as Locales ?? fallbackLng;
	const theme = useContext(ThemeContext);
	const { t } = useTranslation(lang);
	const [isChecked, setIsChecked] = useState<boolean>(theme.value === 'dark');

	const changeThemeHandler: ChangeEventHandler<HTMLInputElement> = () => {
		setIsChecked(!isChecked);
		theme.toggleTheme();
	};

	return (
		<Toggle
			className={className}
			name='theme-toggle'
			id='theme-toggle'
			appearance='secondary'
			labelHidden
			label={t('themeToggle.label')}
			value={theme.value}
			isChecked={isChecked}
			onChange={changeThemeHandler}
			checkedIcon={<MoonIcon />}
			uncheckedIcon={<SunIcon />}
			testId={testId}
			{...props}
		/>
	);
};