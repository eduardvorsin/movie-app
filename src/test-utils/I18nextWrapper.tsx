import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;