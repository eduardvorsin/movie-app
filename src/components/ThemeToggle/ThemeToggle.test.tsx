import { render, screen } from "@testing-library/react";
import ThemeToggle from "./ThemeToggle";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const dictionary = { label: 'Light and dark theme switch' };

describe('ThemeToggle tests', () => {
	it('is rendered correctly', async () => {

		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-theme-toggle')).toBeInTheDocument();
	});

	it('is a basic snapshot', async () => {
		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-theme-toggle')).toMatchSnapshot();
	});
});

describe('ThemeToggle integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('themeToggle.label'));
		const dictionary = { label: i18next.t('themeToggle.label') };
		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByLabelText<HTMLLabelElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('themeToggle.label'));
		const dictionary = { label: i18next.t('themeToggle.label') };
		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByLabelText<HTMLLabelElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');
		const dictionary = { label: i18next.t('themeToggle.label') };
		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-theme-toggle')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');
		const dictionary = { label: i18next.t('themeToggle.label') };
		render(
			<ThemeToggle
				dictionary={dictionary}
				testId='test-theme-toggle'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(await screen.findByTestId<HTMLDivElement>('test-theme-toggle')).toMatchSnapshot();
	});
});