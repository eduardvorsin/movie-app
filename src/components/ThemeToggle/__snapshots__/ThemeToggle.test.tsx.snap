// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle integration tests is a snapshot with English localization 1`] = `
<div
  class="flex items-center undefined"
  data-testid="test-theme-toggle"
>
  <label
    class="text-neutral-800 dark:text-dark-neutral-800 mr-2 transition-colors duration-150  sr-only"
    for="theme-toggle-switch-1"
  >
    Light and dark theme switch
  </label>
  <div
    class="text-neutral-100 dark:text-dark-neutral-100  relative transition-colors duration-150 bg-neutral-800 dark:bg-dark-neutral-800 w-[3.75rem] h-[1.875rem] rounded-[1.875rem] "
  >
    <input
      checked=""
      class="w-full h-full appearance-none cursor-pointer inline-flex p-1 relative disabled:cursor-not-allowed before:bg-neutral-100 dark:before:bg-dark-neutral-100 before:absolute before:rounded-[50%] before:z-100 before:transition-transform before:transition-colors before:duration-150 before:translate-x-0 rounded-[1.875rem] before:w-[1.375rem] before:h-[1.375rem] checked:before:translate-x-[1.875rem]"
      id="theme-toggle-switch-1"
      name="theme-toggle"
      role="switch"
      type="checkbox"
      value="dark"
    />
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] left-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] right-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
  </div>
</div>
`;

exports[`ThemeToggle integration tests is a snapshot with Russian localization 1`] = `
<div
  class="flex items-center undefined"
  data-testid="test-theme-toggle"
>
  <label
    class="text-neutral-800 dark:text-dark-neutral-800 mr-2 transition-colors duration-150  sr-only"
    for="theme-toggle-switch-1"
  >
    Переключатель светлой и темной тем
  </label>
  <div
    class="text-neutral-100 dark:text-dark-neutral-100  relative transition-colors duration-150 bg-neutral-800 dark:bg-dark-neutral-800 w-[3.75rem] h-[1.875rem] rounded-[1.875rem] "
  >
    <input
      checked=""
      class="w-full h-full appearance-none cursor-pointer inline-flex p-1 relative disabled:cursor-not-allowed before:bg-neutral-100 dark:before:bg-dark-neutral-100 before:absolute before:rounded-[50%] before:z-100 before:transition-transform before:transition-colors before:duration-150 before:translate-x-0 rounded-[1.875rem] before:w-[1.375rem] before:h-[1.375rem] checked:before:translate-x-[1.875rem]"
      id="theme-toggle-switch-1"
      name="theme-toggle"
      role="switch"
      type="checkbox"
      value="dark"
    />
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] left-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] right-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
  </div>
</div>
`;

exports[`ThemeToggle tests is a basic snapshot 1`] = `
<div
  class="flex items-center undefined"
  data-testid="test-theme-toggle"
>
  <label
    class="text-neutral-800 dark:text-dark-neutral-800 mr-2 transition-colors duration-150  sr-only"
    for="theme-toggle-switch-1"
  >
    Light and dark theme switch
  </label>
  <div
    class="text-neutral-100 dark:text-dark-neutral-100  relative transition-colors duration-150 bg-neutral-800 dark:bg-dark-neutral-800 w-[3.75rem] h-[1.875rem] rounded-[1.875rem] "
  >
    <input
      checked=""
      class="w-full h-full appearance-none cursor-pointer inline-flex p-1 relative disabled:cursor-not-allowed before:bg-neutral-100 dark:before:bg-dark-neutral-100 before:absolute before:rounded-[50%] before:z-100 before:transition-transform before:transition-colors before:duration-150 before:translate-x-0 rounded-[1.875rem] before:w-[1.375rem] before:h-[1.375rem] checked:before:translate-x-[1.875rem]"
      id="theme-toggle-switch-1"
      name="theme-toggle"
      role="switch"
      type="checkbox"
      value="dark"
    />
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] left-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
    <span
      class="pointer-events-none absolute top-1/2 -translate-y-1/2 z-0 fill-current w-[calc(1.375rem-2px)] h-[calc(1.375rem-2px)] right-[0.3125rem]"
    >
      <test-assets-stub />
    </span>
  </div>
</div>
`;
